spec/parser/groups_spec.rb

```diff
@@ -105,4 +105,15 @@
     expect(root[2][1].number_at_level).to eq 1
     expect(root[2][1][0][1][0].number_at_level).to eq 1
   end
+
+  specify('supports #to_s when frozen') do
+    group_token, literal_token = Regexp::Lexer.lex(/(?:foo)/)
+
+    passive_group =
+      Group::Passive.new(group_token).tap do |expression|
+        expression.expressions = [Literal.new(literal_token)]
+      end.freeze
```

Comment:
So I learned something interesting when trying to set up this regression test: the issue only seems to occur when the group is manually constructed from tokens like this (this has to happen in `mutant` because of how we generate new ASTs and map them back into concrete `Regexp::...` objects and then use those objects to unparse the AST). If you use `RP.parse(/(?:foo)/)` and then call `#freeze` and then `#to_s` that actually works on master. So it seems like something normally initializes the `@implicit` ivar.

It is pretty difficult to manually construct the right regexp objects so this has been tricky for me to debug and figure out how to reproduce the issue.

Reply:
@dgollahon

I was intrigued by this because it seemed to suggest we might inadvertently be calling `#to_s` on expressions while parsing, as that would explain `@implicit` being initialized.

I could not reproduce what you said on master, though. For me, the error occurs with a frozen parse result, too.

spec:

```ruby
  specify('supports #to_s when frozen') do
    group = Regexp::Parser.parse(/(?:foo)/)[0]
    group.freeze
    expect(group.to_s).to eq '(?:foo)'
  end
```

output:

```
Failures:

  1) Group parsing supports #to_s when frozen
     Failure/Error: @implicit ||= false
     
     FrozenError:
       can't modify frozen Regexp::Expression::Group::Passive: ...
```

before you go and change the spec:

I've been thinking about how to have a bit more protection from such problems in the future. My idea is to create a regexp with all or at least a lot of the existing features, parse it (`parse(pattern, :any)`), recursively freeze all expressions, and see if it can be stringified and matches the input. `to_s_spec.rb` would be the place for that. This could actually be a bit of fun, do you want to do it? Otherwise, no problem, I'll merge and release this as is and add work on the spec when I get to it.

Reply:
🤦 ok I think I see what happened here. I think I was freezing the result of of `Regexp::Expression::Root` which is what `RP.parse` returns instead of the first child and somehow never got to the right object to freeze. I tried it several different ways and I think must have just gotten mixed up or was temporarily on the wrong branch or something. Thanks for checking my work here!

I have done a first pass at what I think you have suggested: I updated all the existing assertions in the `to_s_spec` to use deeply frozen parsed objects. To do this easily I added the `ice_nine` gem as a development dependency. If that's a problem / not what you had in mind, let me know! I think that this approach is much better than what I originally had.